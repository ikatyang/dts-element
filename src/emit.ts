import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
