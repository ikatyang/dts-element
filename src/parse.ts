import * as ts from 'typescript';
import { IType } from './collections';
import {
  any_type,
  boolean_type,
  never_type,
  null_type,
  number_type,
  object_type,
  string_type,
  symbol_type,
  this_type,
  undefined_type,
  void_type,
} from './constants';
import { ITypeDeclaration } from './declarations/type-declaration';
import { IElement } from './element';
import { emit_native } from './emit';
import { ITopLevelElement } from './others/top-level-element';
import { parse_array_type } from './parsers/array-type';
import { parse_call_signature } from './parsers/call-signature';
import { parse_class_declaration } from './parsers/class-declaration';
import { parse_construct_signature } from './parsers/construct-signature';
import { parse_constructor } from './parsers/constructor';
import { parse_constructor_type } from './parsers/constructor-type';
import { parse_enum_declaration } from './parsers/enum-declaration';
import { parse_enum_member } from './parsers/enum-member';
import { parse_export_assignment } from './parsers/export-assignment';
import { parse_export_declaration } from './parsers/export-declaration';
import { parse_export_specifier } from './parsers/export-specifier';
import { parse_expression_with_type_arguments } from './parsers/expression-with-type-arguments';
import { parse_function_declaration } from './parsers/function-declaration';
import { parse_function_type } from './parsers/function-type';
import { parse_import_declaration } from './parsers/import-declaration';
import { parse_import_equals_declaration } from './parsers/import-equals-declaration';
import { parse_import_specifier } from './parsers/import-specifier';
import { parse_index_signature } from './parsers/index-signature';
import { parse_indexed_access_type } from './parsers/indexed-access-type';
import { parse_interface_declaration } from './parsers/interface-declaration';
import { parse_intersection_type } from './parsers/intersection-type';
import { parse_mapped_type } from './parsers/mapped-type';
import { parse_method_declaration } from './parsers/method-declaration';
import { parse_method_signature } from './parsers/method-signature';
import { parse_module_declaration } from './parsers/module-declaration';
import { parse_namespace_export_declaration } from './parsers/namespace-export-declaration';
import { parse_parameter_declaration } from './parsers/parameter-declaration';
import { parse_parenthesized_type } from './parsers/parenthesized-type';
import { parse_property_declaration } from './parsers/property-declaration';
import { parse_property_signature } from './parsers/property-signature';
import { parse_source_file } from './parsers/source-file';
import { parse_tuple_type } from './parsers/tuple-type';
import { parse_type_alias_declaration } from './parsers/type-alias-declaration';
import { parse_type_literal } from './parsers/type-literal';
import { parse_type_operator } from './parsers/type-operator';
import { parse_type_parameter } from './parsers/type-parameter';
import { parse_type_predicater } from './parsers/type-predicate';
import { parse_type_query } from './parsers/type-query';
import { parse_type_reference } from './parsers/type-reference';
import { parse_union_type } from './parsers/union-type';
import { parse_variable_statement } from './parsers/variable-statement';
import { create_literal_type } from './types/literal-type';

// tslint:disable:cyclomatic-complexity max-line-length
export const parse_native = (node: ts.Node): IElement<any> => {
  // prettier-ignore
  switch (node.kind) {
    case ts.SyntaxKind.AnyKeyword: return any_type;
    case ts.SyntaxKind.ArrayType: return parse_array_type(node as ts.ArrayTypeNode);
    case ts.SyntaxKind.BooleanKeyword: return boolean_type;
    case ts.SyntaxKind.CallSignature: return parse_call_signature(node as ts.CallSignatureDeclaration);
    case ts.SyntaxKind.ClassDeclaration: return parse_class_declaration(node as ts.ClassDeclaration);
    case ts.SyntaxKind.ConstructSignature: return parse_construct_signature(node as ts.ConstructSignatureDeclaration);
    case ts.SyntaxKind.Constructor: return parse_constructor(node as ts.ConstructorDeclaration);
    case ts.SyntaxKind.ConstructorType: return parse_constructor_type(node as ts.ConstructorTypeNode);
    case ts.SyntaxKind.EnumDeclaration: return parse_enum_declaration(node as ts.EnumDeclaration);
    case ts.SyntaxKind.EnumMember: return parse_enum_member(node as ts.EnumMember);
    case ts.SyntaxKind.ExportAssignment: return parse_export_assignment(node as ts.ExportAssignment);
    case ts.SyntaxKind.ExportDeclaration: return parse_export_declaration(node as ts.ExportDeclaration);
    case ts.SyntaxKind.ExportSpecifier: return parse_export_specifier(node as ts.ExportSpecifier);
    case ts.SyntaxKind.ExpressionWithTypeArguments: return parse_expression_with_type_arguments(node as ts.ExpressionWithTypeArguments);
    case ts.SyntaxKind.FalseKeyword: return create_literal_type({value: false});
    case ts.SyntaxKind.FunctionDeclaration: return parse_function_declaration(node as ts.FunctionDeclaration);
    case ts.SyntaxKind.FunctionType: return parse_function_type(node as ts.FunctionTypeNode);
    case ts.SyntaxKind.ImportDeclaration: return parse_import_declaration(node as ts.ImportDeclaration);
    case ts.SyntaxKind.ImportEqualsDeclaration: return parse_import_equals_declaration(node as ts.ImportEqualsDeclaration);
    case ts.SyntaxKind.ImportSpecifier: return parse_import_specifier(node as ts.ImportSpecifier);
    case ts.SyntaxKind.IndexSignature: return parse_index_signature(node as ts.IndexSignatureDeclaration);
    case ts.SyntaxKind.IndexedAccessType: return parse_indexed_access_type(node as ts.IndexedAccessTypeNode);
    case ts.SyntaxKind.InterfaceDeclaration: return parse_interface_declaration(node as ts.InterfaceDeclaration);
    case ts.SyntaxKind.IntersectionType: return parse_intersection_type(node as ts.IntersectionTypeNode);
    case ts.SyntaxKind.LiteralType: return parse_native((node as ts.LiteralTypeNode).literal);
    case ts.SyntaxKind.MappedType: return parse_mapped_type(node as ts.MappedTypeNode);
    case ts.SyntaxKind.MethodDeclaration: return parse_method_declaration(node as ts.MethodDeclaration);
    case ts.SyntaxKind.MethodSignature: return parse_method_signature(node as ts.MethodSignature);
    case ts.SyntaxKind.ModuleDeclaration: return parse_module_declaration(node as ts.ModuleDeclaration);
    case ts.SyntaxKind.NamespaceExportDeclaration: return parse_namespace_export_declaration(node as ts.NamespaceExportDeclaration);
    case ts.SyntaxKind.NeverKeyword: return never_type;
    case ts.SyntaxKind.NullKeyword: return null_type;
    case ts.SyntaxKind.NumberKeyword: return number_type;
    case ts.SyntaxKind.NumericLiteral: return create_literal_type({value: Number((node as ts.NumericLiteral).text)});
    case ts.SyntaxKind.ObjectKeyword: return object_type;
    case ts.SyntaxKind.Parameter: return parse_parameter_declaration(node as ts.ParameterDeclaration);
    case ts.SyntaxKind.ParenthesizedType: return parse_parenthesized_type(node as ts.ParenthesizedTypeNode);
    case ts.SyntaxKind.PropertyDeclaration: return parse_property_declaration(node as ts.PropertyDeclaration);
    case ts.SyntaxKind.PropertySignature: return parse_property_signature(node as ts.PropertySignature);
    case ts.SyntaxKind.SourceFile: return parse_source_file(node as ts.SourceFile);
    case ts.SyntaxKind.StringKeyword: return string_type;
    case ts.SyntaxKind.StringLiteral: return create_literal_type({value: (node as ts.StringLiteral).text});
    case ts.SyntaxKind.SymbolKeyword: return symbol_type;
    case ts.SyntaxKind.ThisType: return this_type;
    case ts.SyntaxKind.TrueKeyword: return create_literal_type({value: true});
    case ts.SyntaxKind.TupleType: return parse_tuple_type(node as ts.TupleTypeNode);
    case ts.SyntaxKind.TypeAliasDeclaration: return parse_type_alias_declaration(node as ts.TypeAliasDeclaration);
    case ts.SyntaxKind.TypeLiteral: return parse_type_literal(node as ts.TypeLiteralNode);
    case ts.SyntaxKind.TypeOperator: return parse_type_operator(node as ts.TypeOperatorNode);
    case ts.SyntaxKind.TypeParameter: return parse_type_parameter(node as ts.TypeParameterDeclaration);
    case ts.SyntaxKind.TypePredicate: return parse_type_predicater(node as ts.TypePredicateNode);
    case ts.SyntaxKind.TypeQuery: return parse_type_query(node as ts.TypeQueryNode);
    case ts.SyntaxKind.TypeReference: return parse_type_reference(node as ts.TypeReferenceNode);
    case ts.SyntaxKind.UndefinedKeyword: return undefined_type;
    case ts.SyntaxKind.UnionType: return parse_union_type(node as ts.UnionTypeNode);
    case ts.SyntaxKind.VariableStatement: return parse_variable_statement(node as ts.VariableStatement);
    case ts.SyntaxKind.VoidKeyword: return void_type;
    default: throw new Error(`Unexpected ts-kind ${node.kind} ( ${ts.SyntaxKind[node.kind]} ): ${emit_native(node)}`);
  }
};

export const parse = (code: string) =>
  parse_native(
    ts.createSourceFile(
      '',
      code,
      ts.ScriptTarget.Latest,
      /*setParentNodes */ false,
    ),
  ) as ITopLevelElement;

export const parse_type = (type_string: string): IType => {
  const code = `type X = ${type_string}`;
  const top_level_element = parse(code);
  const type_declaration = top_level_element.members[0] as ITypeDeclaration;
  return type_declaration.type!;
};
