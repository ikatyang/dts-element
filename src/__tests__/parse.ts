import * as ts from 'typescript';
import { emit } from '../emit';
import { parse, parse_native, parse_type } from '../parse';

const code = `
/// <reference path="some-path" />
/// <reference types="some-types" />

declare var v_var;
declare let v_let;
declare const v_const;

declare const v_any: any;
declare const v_boolean: boolean;
declare const v_never: never;
declare const v_null: null;
declare const v_number: number;
declare const v_object: object;
declare const v_string: string;
declare const v_symbol: symbol;
declare const v_this: this;
declare const v_undefined: undefined;
declare const v_void: void;

declare const v_array: any[];
declare const v_constructor: new <T>(v: T) => any;
declare const v_function: <T>(v: T) => any;
declare const v_function_predicate_1: () => this is string;
declare const v_function_predicate_2: (v) => v is string;
declare const v_general: X.Y.Z<string>;
declare const v_intersection: string & number;
declare const v_keyof: keyof X;
declare const v_literal_true: true;
declare const v_literal_false: false;
declare const v_literal_number: 123;
declare const v_literal_string: 'str';
declare const v_mapped: {
  [T in keyof X]: any;
};
declare const v_mapped_readonly: {
  readonly [T in keyof X]: any;
};
declare const v_mapped_optional: {
  [T in keyof X]?: any;
};
declare const v_object: {
  readonly a?: any;
  b?<T>(v: T): any;
  o(v?: string): void;
  r(...v: string[]): string;
  r2(...v);
  r3(...v: string[][]);
  new (): any;
  (): any;
  <T>(v: T): T;
  new <T>(v: T): any;
  readonly [index: number]: any;
};
declare const v_sub_1: X[any]['abc'][123];
declare const v_sub_2: A.B.C[X[any]][any];
declare const v_tuple: [string, number];
declare const v_typeof_1: typeof X;
declare const v_typeof_2: typeof X.Y.Z;
declare const v_union: string | number;

import "path/to/somewhere";
import import_default from "path/to/somewhere";
import import_default1, * as import_namespace1 from "path/to/somewhere";
import * as import_namespace2 from "path/to/somewhere";
import import_default2, {im_0 as im_1, im_2} from "path/to/somewhere";
import {im_3 as im_4, im_5} from "path/to/somewhere";

export default X;
export default X.Y.Z;
export = A;
export = A.B.C;
export {ex_0 as ex_1, ex_2};
export as namespace NS;
export * from "path/to/somewhere";
export {ex_3 as ex_4, ex_5} from "path/to/somewhere";

export declare class ZXC extends X.Y.Z {}
export declare abstract class ABC<T, U> extends XYZ<S, V> {
  constructor();
  method?(): void;
  public method1<T>(v: T): T;
  protected method2(): void;
  private method3(): void;
  public abstract method4(): void;
  property?: string;
  public property1: string;
  protected property2: string;
  private property3: string;
  public abstract property4: string;
  "": string;
  ""(): void;
}

export enum Kind {
  A,
  B = 123,
  C,
}

export interface X<T> extends C, D {
  a: T;
  "": T;
  ""(): void;
}

export declare namespace NS {
  const ns: any;
}

declare global {
  interface String {
    toXYZ(): any;
  }
}

declare module "xyz" {
  const aa: string;
}

type TP<T> = T | string;

type adjust = <T, U>(fn: (v: T) => U, index: number, array: T[]) => (T | U)[];

declare function x<T>(v: T): T;

import xyz = require('xyz');

type X = A[B[C]];

type TypeName<T> =
    T extends string ? "string" :
    T extends number ? "number" :
    T extends boolean ? "boolean" :
    T extends undefined ? "undefined" :
    T extends Function ? "function" :
    "object";

type ReturnType<T> = T extends (...args: any[]) => infer R ? R : T;
`;

it('should return correctly', () => {
  expect(emit(parse(code))).toMatchSnapshot();
});

it('should throw error with unexpected kind', () => {
  expect(() => parse_native(ts.createToken(-1))).toThrowError();
});

describe('parse_type', () => {
  it('should return correctly', () => {
    expect(
      parse_type(`
        <T, U>(x: T, y: U) => T | U | void
      `),
    ).toMatchSnapshot();
  });
});
