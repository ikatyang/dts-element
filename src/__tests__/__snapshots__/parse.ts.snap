// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse_type should return correctly 1`] = `
Object {
  "generics": Array [
    Object {
      "defalut": undefined,
      "extends": undefined,
      "kind": 13,
      "name": "T",
    },
    Object {
      "defalut": undefined,
      "extends": undefined,
      "kind": 13,
      "name": "U",
    },
  ],
  "kind": 11,
  "parameters": Array [
    Object {
      "kind": 32,
      "name": "x",
      "optional": undefined,
      "rest": undefined,
      "type": Object {
        "generics": undefined,
        "kind": 12,
        "name": "T",
        "parents": undefined,
      },
    },
    Object {
      "kind": 32,
      "name": "y",
      "optional": undefined,
      "rest": undefined,
      "type": Object {
        "generics": undefined,
        "kind": 12,
        "name": "U",
        "parents": undefined,
      },
    },
  ],
  "return": Object {
    "kind": 41,
    "types": Array [
      Object {
        "generics": undefined,
        "kind": 12,
        "name": "T",
        "parents": undefined,
      },
      Object {
        "generics": undefined,
        "kind": 12,
        "name": "U",
        "parents": undefined,
      },
      Object {
        "kind": 29,
        "type": TokenObject {
          "end": -1,
          "flags": 8,
          "kind": 105,
          "parent": undefined,
          "pos": -1,
        },
      },
    ],
  },
}
`;

exports[`should return correctly 1`] = `
"/// <reference path=\\"some-path\\" />
/// <reference types=\\"some-types\\" />
declare var v_var: any;
declare let v_let: any;
declare const v_const: any;
declare const v_any: any;
declare const v_boolean: boolean;
declare const v_never: never;
declare const v_null: null;
declare const v_number: number;
declare const v_object: object;
declare const v_string: string;
declare const v_symbol: symbol;
declare const v_this: this;
declare const v_undefined: undefined;
declare const v_void: void;
declare const v_array: any[];
declare const v_constructor: new <T>(v: T) => any;
declare const v_function: <T>(v: T) => any;
declare const v_function_predicate_1: () => this is string;
declare const v_function_predicate_2: (v: any) => v is string;
declare const v_general: X.Y.Z<string>;
declare const v_intersection: string & number;
declare const v_keyof: keyof X;
declare const v_literal_true: true;
declare const v_literal_false: false;
declare const v_literal_number: 123;
declare const v_literal_string: \\"str\\";
declare const v_mapped: {
    [T in keyof X]: any;
};
declare const v_mapped_readonly: {
    readonly [T in keyof X]: any;
};
declare const v_mapped_optional: {
    [T in keyof X]?: any;
};
declare const v_object: {
    readonly a?: any;
    b?<T>(v: T): any;
    o(v?: string): void;
    r(...v: string[]): string;
    r2(...v: any[]): any;
    r3(...v: string[][]): any;
    new (): any;
    (): any;
    <T>(v: T): T;
    new <T>(v: T): any;
    readonly [index: number]: any;
};
declare const v_sub_1: X[any][\\"abc\\"][123];
declare const v_sub_2: A.B.C[X[any]][any];
declare const v_tuple: [string, number];
declare const v_typeof_1: typeof X;
declare const v_typeof_2: typeof X.Y.Z;
declare const v_union: string | number;
import \\"path/to/somewhere\\";
import import_default from \\"path/to/somewhere\\";
import import_default1, * as import_namespace1 from \\"path/to/somewhere\\";
import * as import_namespace2 from \\"path/to/somewhere\\";
import import_default2, { im_0 as im_1, im_2 } from \\"path/to/somewhere\\";
import { im_3 as im_4, im_5 } from \\"path/to/somewhere\\";
export default X;
export default X.Y.Z;
export = A;
export = A.B.C;
export { ex_0 as ex_1, ex_2 };
export as namespace NS;
export * from \\"path/to/somewhere\\";
export { ex_3 as ex_4, ex_5 } from \\"path/to/somewhere\\";
export declare class ZXC extends X.Y.Z {
}
export declare abstract class ABC<T, U> extends XYZ<S, V> {
    constructor();
    method?(): void;
    public method1<T>(v: T): T;
    protected method2(): void;
    private method3(): void;
    public abstract method4(): void;
    property?: string;
    public property1: string;
    protected property2: string;
    private property3: string;
    public abstract property4: string;
    \\"\\": string;
    \\"\\"(): void;
}
declare enum Kind {
    A = 0,
    B = 123,
    C = 124
}
export interface X<T> extends C, D {
    a: T;
    \\"\\": T;
    \\"\\"(): void;
}
export declare namespace NS {
    const ns: any;
}
declare global {
    interface String {
        toXYZ(): any;
    }
}
declare module \\"xyz\\" {
    const aa: string;
}
type TP<T> = T | string;
type adjust = <T, U>(fn: (v: T) => U, index: number, array: T[]) => (T | U)[];
declare function x<T>(v: T): T;
import xyz = require(\\"xyz\\");
type X = A[B[C]];
"
`;
